File: cpp/090_SubsetsII.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        std::sort(nums.begin(),nums.end());
        vector<vector<int>> ret;
        vector<int> path;
        backtrace(ret,path,nums,0);
        return ret;
    }
    
private:
    void backtrace(vector<vector<int>>& ret, vector<int>& path, vector<int>& nums, int index){
        ret.push_back(path);
        for(int i = index; i < nums.size(); ++i){
            if(i > index && nums[i] == nums[i-1]) continue;
            path.push_back(nums[i]);
            backtrace(ret,path,nums,i+1);
            path.pop_back();
        }
    }
};

File: js/090_SubsetsII.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function(nums) {
    nums.sort( function(a, b){ return a - b; });
    var ret = [];
    backtrace(ret, [], nums, 0);
    return ret;
};

var backtrace = function(ret, path, nums, index){
    ret.push(path.concat());
    for(var i = index; i < nums.length; ++i){
        if( i > index && nums[i] == nums[i-1] ) continue;
        path.push(nums[i]);
        backtrace(ret, path, nums,i+1);
        path.pop();
    }
};